caching/base.py

```diff
@@ -86,16 +71,39 @@ def query_key(self):
         master), throwing a Django ValueError in the process. Django prevents
         cross DB model saving among related objects.
         """
-        query_db_string = 'qs:%s::db:%s' % (self.query_string, self.db)
+        query_db_string = 'qs:%s::db:%s' % (self.queryset.query_key(), self.db)
         return make_key(query_db_string, with_locale=False)
 
+    def cache_objects(self, objects, query_key):
+        """Cache query_key => objects, then update the flush lists."""
+        log.debug('query_key: %s' % query_key)
+        query_flush = flush_key(self.queryset.query_key())
+        log.debug('query_flush: %s' % query_flush)
+        cache.add(query_key, objects, timeout=self.timeout)
+        invalidator.cache_objects(self.queryset.model, objects, query_key, query_flush)
+
     def __iter__(self):
+        if hasattr(super(CacheInternalCommonMixin, self), '__iter__'):
+            # This is the Django 1.9+ class, so we'll use super().__iter__
+            # which is a ModelIterable iterator.
+            iterator = super(CacheInternalCommonMixin, self).__iter__
+        else:
+            # This is Django 1.8. Use the function passed into the class
+            # constructor.
+            iterator = self.iter_function
+
+        if self.timeout == config.NO_CACHE:
+            # no cache, just iterate and return the results
+            for obj in iterator():
+                yield obj
+            return
+
+        # Try to fetch from the cache.
         try:
             query_key = self.query_key()
         except query.EmptyResultSet:
             raise StopIteration
```

Comment:
Given our findings about `StopIteration` elsewhere, I think this should be converted to a simple `return`